Readme.md
# retext-inspect [![Build Status](https://img.shields.io/travis/wooorm/retext-inspect.svg?style=flat)](https://travis-ci.org/wooorm/retext-inspect) [![Coverage Status](https://img.shields.io/coveralls/wooorm/retext-inspect.svg?style=flat)](https://coveralls.io/r/wooorm/retext-inspect?branch=master)

**[retext](https://github.com/wooorm/retext "Retext")** node inspector.

## Installation

[npm](https://docs.npmjs.com/cli/install):

```bash
$ npm install retext-inspect
```

[Component.js](https://github.com/componentjs/component):

```bash
$ component install wooorm/retext-inspect
```

[Bower](http://bower.io/#install-packages):

```bash
$ bower install retext-inspect
```

[Duo](http://duojs.org/#getting-started):

```javascript
var inspect = require('wooorm/retext-inspect');
```

## Usage

```javascript
var Retext = require('retext');
var inspect = require('retext-inspect');

var retext = new Retext().use(inspect);

retext.parse('Some simple text.', function (err, tree) {
    /* Inspect a node: */
    console.log(tree.head.head.tail.inspect());
    /*
     * Logs:
     *
     * PunctuationNode[1]
     * └─ TextNode: '.'
     */

    /* Inspect a node in Node.js: */
    console.log(tree);
    /*
     * Logs:
     *
     * RootNode[1]
     * └─ ParagraphNode[1]
     *    └─ SentenceNode[6]
     *       ├─ WordNode[1]
     *       │  └─ TextNode: 'Some'
     *       ├─ WhiteSpaceNode[1]
     *       │  └─ TextNode: ' '
     *       ├─ WordNode[1]
     *       │  └─ TextNode: 'simple'
     *       ├─ WhiteSpaceNode[1]
     *       │  └─ TextNode: ' '
     *       ├─ WordNode[1]
     *       │  └─ TextNode: 'text'
     *       └─ PunctuationNode[1]
     *          └─ TextNode: '.'
     */
});
```

## API

### [Retext#use](https://github.com/wooorm/retext#retextuseplugin-options)(inspect, options?)

```javascript
retext = new Retext().use(inspect, {
    'color': false
});
```

Options:

- `color` (`boolean`, default `true`) - The diagram returned by `inspect` uses color to make things more awesome. Support is detected by checking if a `util.inspect()` exists (it does in Node.js). To overwrite this, pass an `options` object as a second argument to `Retext#use()`, with a `color` property set to `false`.

### [Node](https://github.com/wooorm/textom#textomnode-nlcstnode)#inspect()

Get a string representing `node`.

There’s no need to call `Node#inspect()` when in Node.js. In Node.js, `console.log` and `util.inspect` all [show the tree diagram](http://nodejs.org/api/util.html#util_util_inspect_object_options) as seen above.

To log the nodes as normal objects, pass `customInspect: false` to `util.inspect()`:

```javascript
console.log(util.inspect(tree, {
    'customInspect': false
}))
```

## Performance

```text
             TextOM.Node#inspect()
  1,062 op/s » A paragraph (5 sentences, 100 words)
    101 op/s » A section (10 paragraphs, 50 sentences, 1,000 words)
      9 op/s » An article (100 paragraphs, 500 sentences, 10,000 words)
```

## License

[MIT](LICENSE) © [Titus Wormer](http://wooorm.com)


History.md

0.4.2 / 2015-01-13
==================

  * Update copyright notice in `LICENSE` to include 2015
  * Add link to whole license in `Readme.md`
  * Add Duo as an instalation method in `Readme.md`
  * Add links to installation methods in `Readme.md`
  * Refactor API reference in `Readme.md`
  * Refactor code examples in `Readme.md`
  * Refactor fences code blocks in `Readme.md`
  * Refactor to adhere to strict jsdoc style
  * Add jscs-jsdoc configuration to `.jscs.json`
  * Add jscs-jsdoc as a dev-dependency
  * Refactor npm scripts for changes in npm
  * Update eslint
  * Update retext

0.4.1 / 2014-12-06
==================

 * Add benchmark
 * Refactor npm script targets in `package.json`
 * Add link to personal website to copyright in `Readme.md`
 * Update eslint, retext

0.4.0 / 2014-11-20
==================

 * Add support for data attributes
 * Refactor useage examples in `Readme.md`
 * Fix incorrect executive rights on `test.js`
 * Add flat badges to `Readme.md`
 * Refactor npm script targets in `package.json`
 * Update retext

0.3.0 / 2014-10-28
==================

 * Refactor for standard retext options

0.2.2 / 2014-10-27
==================

 * Add `.eslintrc`
 * Refactor to disallow space after object keys
 * Refactor for changes in retext
 * Update eslint, mocha, retext

0.2.1 / 2014-10-21
==================

 * Refactor test
 * Refactor test for changes in retext
 * Update retext

0.2.0 / 2014-10-14
==================

 * Remove extraneous API options
 * Update .gitignore, .npmignore, bower ignore
 * Move spec to test.js

0.1.0 / 2014-10-06
==================

 * Add support for working with console.log and util.inspect


test.js
'use strict';

var inspect,
    chalk,
    Retext,
    assert,
    util;

/*
 * Module dependencies.
 */

inspect = require('./');
chalk = require('chalk');
Retext = require('retext');
assert = require('assert');
util = require('util');

/**
 * Intercept stdout.
 *
 * @param {function(string)} callback
 * @return {Function} A method to stop intercepting.
 */
function intercept(callback) {
    var write;

    write = process.stdout.write;

    process.stdout.write = callback;

    return function () {
        process.stdout.write = write;
    };
}

/*
 * Fixtures.
 */

var paragraph;

paragraph = 'Some simple text. Other “sentence”.';

/*
 * Retext.
 */

var retextWithColor,
    retextWithoutColor,
    TextOM;

retextWithColor = new Retext().use(inspect);
retextWithoutColor = new Retext().use(inspect, {
    'color': false
});

TextOM = retextWithColor.TextOM;

/*
 * Tests.
 */

describe('retext-inspect()', function () {
    it('should be a `function`', function () {
        assert(typeof inspect === 'function');
    });

    it('should attach an `inspect` method to `Node#`', function () {
        assert(typeof (new TextOM.Parent()).inspect === 'function');
        assert(typeof (new TextOM.Child()).inspect === 'function');
        assert(typeof (new TextOM.Element()).inspect === 'function');
        assert(typeof (new TextOM.Text()).inspect === 'function');
        assert(typeof (new TextOM.RootNode()).inspect === 'function');
        assert(typeof (new TextOM.ParagraphNode()).inspect === 'function');
        assert(typeof (new TextOM.SentenceNode()).inspect === 'function');
        assert(typeof (new TextOM.WordNode()).inspect === 'function');
        assert(typeof (new TextOM.SymbolNode()).inspect === 'function');
        assert(typeof (new TextOM.PunctuationNode()).inspect === 'function');
        assert(typeof (new TextOM.WhiteSpaceNode()).inspect === 'function');
        assert(typeof (new TextOM.TextNode()).inspect === 'function');
        assert(typeof (new TextOM.SourceNode()).inspect === 'function');
    });
});

/*
 * Unit tests for `Node#inspect()`.
 */

describe('inspect()', function () {
    var tree;

    before(function (done) {
        retextWithColor.parse(paragraph, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should work on `RootNode`', function () {
        var fixture;

        fixture = [
            'RootNode[1]',
            '└─ ParagraphNode[3]',
            '   ├─ SentenceNode[6]',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'Some\'',
            '   │  ├─ WhiteSpaceNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'simple\'',
            '   │  ├─ WhiteSpaceNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'text\'',
            '   │  └─ PunctuationNode: \'.\'',
            '   ├─ WhiteSpaceNode: \' \'',
            '   └─ SentenceNode[6]',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'Other\'',
            '      ├─ WhiteSpaceNode: \' \'',
            '      ├─ PunctuationNode: \'“\'',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'sentence\'',
            '      ├─ PunctuationNode: \'”\'',
            '      └─ PunctuationNode: \'.\''
        ].join('\n');

        assert(chalk.stripColor(tree.inspect()) === fixture);
    });

    it('should work on `SentenceNode`', function () {
        var fixture;

        fixture = [
            'SentenceNode[6]',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'Some\'',
            '├─ WhiteSpaceNode: \' \'',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'simple\'',
            '├─ WhiteSpaceNode: \' \'',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'text\'',
            '└─ PunctuationNode: \'.\''
        ].join('\n');

        assert(chalk.stripColor(tree.head.head.inspect()) === fixture);
    });

    it('should work with data attributes', function () {
        var text,
            fixture;

        text = new TextOM.SymbolNode('$');

        text.data.test = true;

        fixture = 'SymbolNode: \'$\' [data={"test":true}]';

        assert(chalk.stripColor(text.inspect()) === fixture);
    });
});

describe('use(inspect, {color: false})', function () {
    var sentence;

    before(function (done) {
        retextWithoutColor.parse(paragraph, function (err, tree) {
            sentence = tree.head.head;

            done(err);
        });
    });

    it('should work', function () {
        var fixture;

        fixture = [
            'SentenceNode[6]',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'Some\'',
            '├─ WhiteSpaceNode: \' \'',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'simple\'',
            '├─ WhiteSpaceNode: \' \'',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'text\'',
            '└─ PunctuationNode: \'.\''
        ].join('\n');

        assert(sentence.inspect() === fixture);
    });
});

describe('use(inspect, {color: true})', function () {
    var sentence;

    before(function (done) {
        retextWithColor.parse(paragraph, function (err, tree) {
            sentence = tree.head.head;

            done(err);
        });
    });

    it('should work', function () {
        var fixture;

        fixture = [
            'SentenceNode' +
                chalk.dim('[') + chalk.yellow('6') + chalk.dim(']'),
            chalk.dim('├─ ') + 'WordNode' +
                chalk.dim('[') + chalk.yellow('1') + chalk.dim(']'),
            chalk.dim('│  └─ ') + 'TextNode' +
                chalk.dim(': \'') + chalk.green('Some') + chalk.dim('\''),
            chalk.dim('├─ ') + 'WhiteSpaceNode' +
                chalk.dim(': \'') + chalk.green(' ') + chalk.dim('\''),
            chalk.dim('├─ ') + 'WordNode' +
                chalk.dim('[') + chalk.yellow('1') + chalk.dim(']'),
            chalk.dim('│  └─ ') + 'TextNode' +
                chalk.dim(': \'') + chalk.green('simple') + chalk.dim('\''),
            chalk.dim('├─ ') + 'WhiteSpaceNode' +
                chalk.dim(': \'') + chalk.green(' ') + chalk.dim('\''),
            chalk.dim('├─ ') + 'WordNode' +
                chalk.dim('[') + chalk.yellow('1') + chalk.dim(']'),
            chalk.dim('│  └─ ') + 'TextNode' +
                chalk.dim(': \'') + chalk.green('text') + chalk.dim('\''),
            chalk.dim('└─ ') + 'PunctuationNode' +
                chalk.dim(': \'') + chalk.green('.') + chalk.dim('\'')
        ].join('\n');

        assert(sentence.inspect() === fixture);
    });
});

/*
 * Unit tests for Node.js integration.
 */

describe('`util.inspect` and `console.log` integration', function () {
    var tree;

    before(function (done) {
        retextWithColor.parse(paragraph, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should work with `util.inspect`', function () {
        var fixture;

        fixture = [
            'RootNode[1]',
            '└─ ParagraphNode[3]',
            '   ├─ SentenceNode[6]',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'Some\'',
            '   │  ├─ WhiteSpaceNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'simple\'',
            '   │  ├─ WhiteSpaceNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'text\'',
            '   │  └─ PunctuationNode: \'.\'',
            '   ├─ WhiteSpaceNode: \' \'',
            '   └─ SentenceNode[6]',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'Other\'',
            '      ├─ WhiteSpaceNode: \' \'',
            '      ├─ PunctuationNode: \'“\'',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'sentence\'',
            '      ├─ PunctuationNode: \'”\'',
            '      └─ PunctuationNode: \'.\''
        ].join('\n');

        assert(chalk.stripColor(util.inspect(tree)) === fixture);
    });

    it('should work with `console.log`', function (done) {
        var fixture,
            stop;

        fixture = [
            'RootNode[1]',
            '└─ ParagraphNode[3]',
            '   ├─ SentenceNode[6]',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'Some\'',
            '   │  ├─ WhiteSpaceNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'simple\'',
            '   │  ├─ WhiteSpaceNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'text\'',
            '   │  └─ PunctuationNode: \'.\'',
            '   ├─ WhiteSpaceNode: \' \'',
            '   └─ SentenceNode[6]',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'Other\'',
            '      ├─ WhiteSpaceNode: \' \'',
            '      ├─ PunctuationNode: \'“\'',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'sentence\'',
            '      ├─ PunctuationNode: \'”\'',
            '      └─ PunctuationNode: \'.\''
        ].join('\n');

        stop = intercept(function (value) {
            stop();

            assert(chalk.stripColor(value) === fixture + '\n');

            done();
        });

        console.log(tree);
    });
});


spec/retext-inspect.spec.js
'use strict';

var inspect,
    chalk,
    Retext,
    assert,
    util,
    retext,
    TextOM,
    paragraph;

/**
 * Module dependencies.
 */

inspect = require('..');
Retext = require('retext');
chalk = require('chalk');
assert = require('assert');
util = require('util');

/**
 * Intercept stdout.
 *
 * @param {function(string)} callback
 * @return {Function} A method to stop intercepting.
 */

function intercept(callback) {
    var write;

    write = process.stdout.write;

    process.stdout.write = callback;

    return function () {
        process.stdout.write = write;
    };
}

/**
 * Format nesting.
 *
 * @param {string} nesting
 * @return {string}
 */

function formatNesting(nesting) {
    return chalk.dim(nesting);
}

/**
 * Format node.
 *
 * @param {Node} node
 * @return {string}
 */

function formatNode(node) {
    if ('length' in node) {
        return node.type + '[' + chalk.yellow(node.length) + ']';
    }

    return node.type + ': ' + chalk.green('\'' + node.toString() + '\'');
}

/**
 * Constants.
 */

paragraph = 'Some simple text. Other “sentence”.';

retext = new Retext().use(inspect);
TextOM = retext.TextOM;

/**
 * Unit tests.
 */

describe('retext-inspect()', function () {
    it('should be a `function`', function () {
        assert(typeof inspect === 'function');
    });

    it('should attach an `inspect` method to `Node#`', function () {
        assert(typeof (new TextOM.Parent()).inspect === 'function');
        assert(typeof (new TextOM.Child()).inspect === 'function');
        assert(typeof (new TextOM.Element()).inspect === 'function');
        assert(typeof (new TextOM.Text()).inspect === 'function');
        assert(typeof (new TextOM.RootNode()).inspect === 'function');
        assert(typeof (new TextOM.ParagraphNode()).inspect === 'function');
        assert(typeof (new TextOM.SentenceNode()).inspect === 'function');
        assert(typeof (new TextOM.WordNode()).inspect === 'function');
        assert(typeof (new TextOM.WhiteSpaceNode()).inspect === 'function');
        assert(typeof (new TextOM.PunctuationNode()).inspect === 'function');
        assert(typeof (new TextOM.TextNode()).inspect === 'function');
        assert(typeof (new TextOM.SourceNode()).inspect === 'function');
    });
});

/**
 * Unit tests for `Node#inspect()`.
 */

describe('inspect()', function () {
    var tree;

    before(function (done) {
        retext.parse(paragraph, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should work on `RootNode`', function () {
        var fixture;

        fixture = [
            'RootNode[1]',
            '└─ ParagraphNode[3]',
            '   ├─ SentenceNode[6]',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'Some\'',
            '   │  ├─ WhiteSpaceNode[1]',
            '   │  │  └─ TextNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'simple\'',
            '   │  ├─ WhiteSpaceNode[1]',
            '   │  │  └─ TextNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'text\'',
            '   │  └─ PunctuationNode[1]',
            '   │     └─ TextNode: \'.\'',
            '   ├─ WhiteSpaceNode[1]',
            '   │  └─ TextNode: \' \'',
            '   └─ SentenceNode[6]',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'Other\'',
            '      ├─ WhiteSpaceNode[1]',
            '      │  └─ TextNode: \' \'',
            '      ├─ PunctuationNode[1]',
            '      │  └─ TextNode: \'“\'',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'sentence\'',
            '      ├─ PunctuationNode[1]',
            '      │  └─ TextNode: \'”\'',
            '      └─ PunctuationNode[1]',
            '         └─ TextNode: \'.\''
        ].join('\n');

        assert(tree.inspect() === fixture);
    });

    it('should work on `SentenceNode`', function () {
        var fixture;

        fixture = [
            'SentenceNode[6]',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'Some\'',
            '├─ WhiteSpaceNode[1]',
            '│  └─ TextNode: \' \'',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'simple\'',
            '├─ WhiteSpaceNode[1]',
            '│  └─ TextNode: \' \'',
            '├─ WordNode[1]',
            '│  └─ TextNode: \'text\'',
            '└─ PunctuationNode[1]',
            '   └─ TextNode: \'.\''
        ].join('\n');

        assert(tree.head.head.inspect() === fixture);
    });
});

describe('inspect({formatNesting: function(nesting)})', function () {
    var sentence;

    before(function (done) {
        retext.parse(paragraph, function (err, tree) {
            sentence = tree.head.head;

            done(err);
        });
    });

    it('should work', function () {
        var fixture;

        fixture = [
            'SentenceNode[6]',
            chalk.dim('├─ ') + 'WordNode[1]',
            chalk.dim('│  └─ ') + 'TextNode: \'Some\'',
            chalk.dim('├─ ') + 'WhiteSpaceNode[1]',
            chalk.dim('│  └─ ') + 'TextNode: \' \'',
            chalk.dim('├─ ') + 'WordNode[1]',
            chalk.dim('│  └─ ') + 'TextNode: \'simple\'',
            chalk.dim('├─ ') + 'WhiteSpaceNode[1]',
            chalk.dim('│  └─ ') + 'TextNode: \' \'',
            chalk.dim('├─ ') + 'WordNode[1]',
            chalk.dim('│  └─ ') + 'TextNode: \'text\'',
            chalk.dim('└─ ') + 'PunctuationNode[1]',
            chalk.dim('   └─ ') + 'TextNode: \'.\''
        ].join('\n');

        assert(sentence.inspect({
            'formatNesting' : formatNesting
        }) === fixture);
    });
});

describe('inspect({formatNode: function(node)})', function () {
    var sentence;

    before(function (done) {
        retext.parse(paragraph, function (err, tree) {
            sentence = tree.head.head;

            done(err);
        });
    });

    it('should work', function () {
        var fixture;

        fixture = [
            'SentenceNode[' + chalk.yellow(6) + ']',
            '├─ WordNode[' + chalk.yellow(1) + ']',
            '│  └─ TextNode: ' + chalk.green('\'Some\''),
            '├─ WhiteSpaceNode[' + chalk.yellow(1) + ']',
            '│  └─ TextNode: ' + chalk.green('\' \''),
            '├─ WordNode[' + chalk.yellow(1) + ']',
            '│  └─ TextNode: ' + chalk.green('\'simple\''),
            '├─ WhiteSpaceNode[' + chalk.yellow(1) + ']',
            '│  └─ TextNode: ' + chalk.green('\' \''),
            '├─ WordNode[' + chalk.yellow(1) + ']',
            '│  └─ TextNode: ' + chalk.green('\'text\''),
            '└─ PunctuationNode[' + chalk.yellow(1) + ']',
            '   └─ TextNode: ' + chalk.green('\'.\'')
        ].join('\n');

        assert(sentence.inspect({
            'formatNode' : formatNode
        }) === fixture);
    });
});

/**
 * Unit tests for Node.js integration.
 */

describe('`util.inspect` and `console.log` integration', function () {
    var tree;

    before(function (done) {
        retext.parse(paragraph, function (err, node) {
            tree = node;

            done(err);
        });
    });

    it('should work with `util.inspect`', function () {
        var fixture;

        fixture = [
            'RootNode[1]',
            '└─ ParagraphNode[3]',
            '   ├─ SentenceNode[6]',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'Some\'',
            '   │  ├─ WhiteSpaceNode[1]',
            '   │  │  └─ TextNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'simple\'',
            '   │  ├─ WhiteSpaceNode[1]',
            '   │  │  └─ TextNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'text\'',
            '   │  └─ PunctuationNode[1]',
            '   │     └─ TextNode: \'.\'',
            '   ├─ WhiteSpaceNode[1]',
            '   │  └─ TextNode: \' \'',
            '   └─ SentenceNode[6]',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'Other\'',
            '      ├─ WhiteSpaceNode[1]',
            '      │  └─ TextNode: \' \'',
            '      ├─ PunctuationNode[1]',
            '      │  └─ TextNode: \'“\'',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'sentence\'',
            '      ├─ PunctuationNode[1]',
            '      │  └─ TextNode: \'”\'',
            '      └─ PunctuationNode[1]',
            '         └─ TextNode: \'.\''
        ].join('\n');

        assert(util.inspect(tree) === fixture);
    });

    it('should work with `console.log`', function (done) {
        var fixture,
            stop;

        fixture = [
            'RootNode[1]',
            '└─ ParagraphNode[3]',
            '   ├─ SentenceNode[6]',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'Some\'',
            '   │  ├─ WhiteSpaceNode[1]',
            '   │  │  └─ TextNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'simple\'',
            '   │  ├─ WhiteSpaceNode[1]',
            '   │  │  └─ TextNode: \' \'',
            '   │  ├─ WordNode[1]',
            '   │  │  └─ TextNode: \'text\'',
            '   │  └─ PunctuationNode[1]',
            '   │     └─ TextNode: \'.\'',
            '   ├─ WhiteSpaceNode[1]',
            '   │  └─ TextNode: \' \'',
            '   └─ SentenceNode[6]',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'Other\'',
            '      ├─ WhiteSpaceNode[1]',
            '      │  └─ TextNode: \' \'',
            '      ├─ PunctuationNode[1]',
            '      │  └─ TextNode: \'“\'',
            '      ├─ WordNode[1]',
            '      │  └─ TextNode: \'sentence\'',
            '      ├─ PunctuationNode[1]',
            '      │  └─ TextNode: \'”\'',
            '      └─ PunctuationNode[1]',
            '         └─ TextNode: \'.\''
        ].join('\n');

        stop = intercept(function (value) {
            stop();

            assert(value === fixture + '\n');

            done();
        });

        console.log(tree);
    });
});


index.js
'use strict';

/*
 * Cached methods.
 */

var has;

has = Object.prototype.hasOwnProperty;

/**
 * Get whether `object` has keys.
 *
 * @param {Object} object
 * @return {boolean}
 */
function hasKeys(object) {
    var key;

    for (key in object) {
        /* istanbul ignore else */
        if (has.call(object, key)) {
            return true;
        }
    }

    return false;
}

/*
 * Define ANSII color functions.
 */

var dim,
    yellow,
    green;

/**
 * Factory to wrap values in ANSI colours.
 *
 * @param {string} open
 * @param {string} close
 * @return {function(string): string}
 */
function ansiColor(open, close) {
    return function (value) {
        return '\u001b[' + open + 'm' + value + '\u001b[' + close + 'm';
    };
}

dim = ansiColor(2, 22);
yellow = ansiColor(33, 39);
green = ansiColor(32, 39);

/*
 * Define ANSII color removal functionality.
 */

var COLOR_EXPRESSION;

COLOR_EXPRESSION = new RegExp(
    '(?:' +
        '(?:\\u001b\\[)|' +
        '\\u009b' +
    ')' +
    '(?:' +
        '(?:[0-9]{1,3})?(?:(?:;[0-9]{0,3})*)?[A-M|f-m]' +
    ')|' +
    '\\u001b[A-M]',
    'g'
);

/**
 * Remove ANSI colour from `value`.
 *
 * @param {string} value
 * @return {string}
 */
function stripColor(value) {
    return value.replace(COLOR_EXPRESSION, '');
}

/*
 * Define constants characters.
 */

var CHAR_VERTICAL_LINE,
    CHAR_HORIZONTAL_LINE,
    CHAR_SPLIT,
    CHAR_CONTINUE_AND_SPLIT,
    CONTINUE,
    STOP;

CHAR_HORIZONTAL_LINE = '─';
CHAR_VERTICAL_LINE = '│';
CHAR_SPLIT = '└';
CHAR_CONTINUE_AND_SPLIT = '├';

CONTINUE = CHAR_CONTINUE_AND_SPLIT + CHAR_HORIZONTAL_LINE + ' ';
STOP = CHAR_SPLIT + CHAR_HORIZONTAL_LINE + ' ';

/**
 * Colored nesting formatter.
 *
 * @param {string} value
 * @return {string}
 */
function formatNesting(value) {
    return dim(value);
}

/**
 * Colored node formatter.
 *
 * @param {Node} node
 * @return {string}
 */
function formatNode(node) {
    var log;

    log = node.type;

    if ('length' in node) {
        log += dim('[') + yellow(node.length) + dim(']');
    } else {
        log += dim(': \'') + green(node.toString()) + dim('\'');
    }

    if (hasKeys(node.data)) {
        log += ' [data=' + JSON.stringify(node.data) + ']';
    }

    return log;
}

/**
 * Inspects a node.
 *
 * @this {Node}
 * @return {string}
 */
function inspect(pad) {
    var self,
        node,
        result,
        index,
        length;

    self = this;

    if (!('length' in self)) {
        return formatNode(self);
    }

    if (!pad || typeof pad === 'number') {
        pad = '';
    }

    result = [];

    result.push(formatNode(self));

    index = -1;
    length = self.length;

    while (++index < length) {
        node = self[index];

        if (index === length - 1) {
            result.push(
                formatNesting(pad + STOP) + node.inspect(pad + '   ')
            );
        } else {
            result.push(
                formatNesting(pad + CONTINUE) +
                    node.inspect(pad + CHAR_VERTICAL_LINE + '  ')
            );
        }
    }

    return result.join('\n');
}

/**
 * Inspects a node, without using color.
 *
 * @return {string}
 * @this {Node}
 */
function inspectWithoutColor(pad) {
    return stripColor(inspect.call(this, pad));
}

/**
 * Define `plugin`.
 *
 * @param {Retext} retext
 */
function plugin(retext, options) {
    var color;

    color = options.color;

    if (typeof color !== 'boolean') {
        /*
         * Detect if the client has a native `util.inspect`.
         * If so, turn `color` on.
         */

        try {
            color = 'inspect' in require('util');
        } catch (exception) {
            /* istanbul ignore next */
            color = false;
        }
    }

    /*
     * Expose `inspect` on `Node`s.
     */

    retext.TextOM.Node.prototype.inspect = color ?
        inspect : inspectWithoutColor;
}

/*
 * Expose `plugin`.
 */

exports = module.exports = plugin;


benchmark.js
'use strict';

var Retext,
    inspect;

/*
 * Dependencies.
 */

Retext = require('retext');
inspect = require('./');

/*
 * Dependencies.
 */

var retext;

retext = new Retext().use(inspect);

/*
 * Test data.
 *
 * This includes:
 *
 * - An average sentence (w/ 20 words);
 * - An average paragraph (w/ 5 sentences);
 * - A (big?) section (w/ 10 paragraphs);
 * - A (big?) article (w/ 10 sections);
 *
 * Source:
 *   http://www.gutenberg.org/files/10745/10745-h/10745-h.htm
 */

var sentence,
    paragraph,
    section,
    article;

sentence = 'Where she had stood was clear, and she was gone since Sir ' +
    'Kay does not choose to assume my quarrel.';

paragraph = 'Thou art a churlish knight to so affront a lady ' +
    'he could not sit upon his horse any longer. ' +
    'For methinks something hath befallen my lord and that he ' +
    'then, after a while, he cried out in great voice. ' +
    'For that light in the sky lieth in the south ' +
    'then Queen Helen fell down in a swoon, and lay. ' +
    'Touch me not, for I am not mortal, but Fay ' +
    'so the Lady of the Lake vanished away, everything behind. ' +
    sentence;

section = paragraph + Array(10).join('\n\n' + paragraph);

article = section + Array(10).join('\n\n' + section);

before(function (done) {
    retext.parse(paragraph, function (err, tree) {
        paragraph = tree;

        done(err);
    });
});

before(function (done) {
    retext.parse(section, function (err, tree) {
        section = tree;

        done(err);
    });
});

before(function (done) {
    retext.parse(article, function (err, tree) {
        article = tree;

        done(err);
    });
});

suite('TextOM.Node#inspect()', function () {
    bench('A paragraph (5 sentences, 100 words)', function () {
        paragraph.inspect();
    });

    bench('A section (10 paragraphs, 50 sentences, 1,000 words)',
        function () {
            section.inspect();
        }
    );

    bench('An article (100 paragraphs, 500 sentences, 10,000 words)',
        function () {
            article.inspect();
        }
    );
});
